### DISCUSSION TITLE
Wifi not connecting

### DISCUSSION THREAD

**Discussion by xvlvx0** [micropython/micropython#10375] [ESP32]
Hi All,

I'm having a hard time getting the wifi to connect in the production (home use) environment, on a Wemos S2 mini (ESP32).

It works fine on the desk test setup, it maybe takes a few seconds but most of times without any hiccups. Just not when connected to the production wires. The 4 wires that are connected are: GND, 5V, Rx (pin 9) and Tx (Pin 10). These wires (UART) connect via a 2m cable to a UART appliance. 

The following function is run when the board boots, and every xx seconds.

`
def setup_wifi() -> None:
    """Set up the wifi connection."""
    print("\nsetup_wifi()")

    global wlan
    global wifi_settings

    wifi_settings["networkAvailable"] = False
    wlan = network.WLAN(network.STA_IF)  # create station interface
    wlan.active(True)  # activate the interface
    time.sleep(0.5)

    if not wlan.isconnected():
        print("connecting to wifi network...")

        wlan.connect(wifi_settings["ssid"], wifi_settings["passwd"])  # connect to an AP
        time.sleep(0.5)

        retry_count = 0
        while not wlan.isconnected():
            retry_count += 1
            print("..wifi nope.." + str(retry_count) + "/5")
            time.sleep(3)

            if retry_count >= 5:
                time.sleep(10)
                print("reset machine")
                time.sleep(1)
                machine.reset()

        print("...wifi connected.")
    else:
        print("Wifi already connected.")

    wifi_settings["networkAvailable"] = True
    ip, mask, gateway, dns = wlan.ifconfig()  # get the interface's IP/netmask/gw/DNS addresses
    wifi_settings["ip_address"] = ip
    print("wifi settings, IP: " + ip)
`

As dictated by the code, if it cannot connect to the wifi it retries a maximum of 5 times and stops and resets the board. I choose to do it this way to recover if the wifi connection got lost and a reconnect isn't working out. The code is called for ever so often, making sure the Wemos S2 keeps its connection.

Does anyone have had similar problems? Is it due to the length of the UART cable?

Best regards,
lvx.

P.S. just to test I switched the cable for a double twisted pair cable (2 pairs of a 4 pair UTP cable), hoping it would remedy the issue, sadly it didn't!

**Comment by robert-hh** [comment]
The wire length for UART is not critical, and UART and WiFi should not interfere at all. Do you get print output at the UART? Just be sure that the signal level at RX/TX is 3.3V, not 5V. But even that should not interfere with WiFi.

**Comment by davefes** [comment]
Try throwing a try/except around connect():
```py
        try:
             sta_if.connect(config.hotspot, config.password)
        except OSError as error:
            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write(str(error) + '\n')
            except OSError:
                pass
```
Or print statements and try to determine how far through the def setup_wifi() process you get through.
I have a 1000uF low-esr electro on the supply feed from the FTDI programmer as there can be drop-outs on the supply when the WiFi block is enabled.

**Reply by davefes** [reply]
Also, I use a:
```py
#  seems more reliable to start with a fresh connect()
    if sta_if.isconnected():
        sta_if.disconnect()
        print (f'started in the connected state, but now disconnected')
    else:
        print (f'started in the disconnected state')
```
after the wlan.active()

**Comment by PBrunot** [comment]
You can try this:
* The WEMOS C3 mini required reduced TX power from Wifi to workaround some board issue (antenna design)
* See https://www.wemos.cc/en/latest/c3/c3_mini_1_0_0.html

```python
      sta_if.active(True)
      sta_if.config(txpower=8.5)
      sta_if.connect('ssid', 'passwd')
```
